// nexmo/lib/Conversation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _index = require('./index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Conversations = function () {
  _createClass(Conversations, null, [{
    key: 'PATH',
    get: function get() {
      return '/beta/conversations';
    }

    /**
     * @param {Credentials} credentials
     *    credentials to be used when interacting with the API.
     * @param {Object} options
     *    Additional Conversations options.
     */

  }]);

  function Conversations(credentials) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, Conversations);

    this.creds = credentials;
    this.options = options;

    // Used to facilitate testing of the call to the underlying object
    this._nexmo = this.options.nexmoOverride || _index2.default;
  }

  /**
   * TODO: document
   */


  _createClass(Conversations, [{
    key: 'create',
    value: function create(params, callback) {
      params = JSON.stringify(params);

      var config = {
        host: 'api.nexmo.com',
        path: Conversations.PATH,
        method: 'POST',
        body: params,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer ' + this.creds.generateJwt()
        }
      };
      this.options.httpClient.request(config, callback);
    }

    /**
     * TODO: document
     */

  }, {
    key: 'get',
    value: function get(conversationId, callback) {
      var config;
      if (conversationId) {
        config = {
          host: 'api.nexmo.com',
          path: Conversations.PATH + '/' + conversationId,
          method: 'GET',
          headers: {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer ' + this.creds.generateJwt()
          }
        };
      } else {
        config = {
          host: 'api.nexmo.com',
          path: Conversations.PATH,
          method: 'GET',
          headers: {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer ' + this.creds.generateJwt()
          }
        };
      }

      this.options.httpClient.request(config, callback);
    }

    /**
     * TODO: document
     */

  }, {
    key: 'update',
    value: function update(conversationId, params, callback) {
      params = JSON.stringify(params);

      var config = {
        host: 'api.nexmo.com',
        path: Conversations.PATH + '/' + conversationId,
        method: 'PUT',
        body: params,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer ' + this.creds.generateJwt()
        }
      };

      this.options.httpClient.request(config, callback);
    }

    /**
     * TODO: document
     */

  }, {
    key: 'delete',
    value: function _delete(conversationId, callback) {
      var config = {
        host: 'api.nexmo.com',
        path: Conversations.PATH + '/' + conversationId,
        method: 'DELETE',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer ' + this.creds.generateJwt()
        }
      };

      this.options.httpClient.request(config, callback);
    }
  }]);

  return Conversations;
}();

exports.default = Conversations;
module.exports = exports['default'];
